libs/langgraph/src/agent/agent.ts

```diff
@@ -0,0 +1,203 @@
+}
+
+function isBaseChatModel(model: LanguageModelLike): model is BaseChatModel {
+  return (
+    "invoke" in model &&
+    typeof model.invoke === "function" &&
+    "_modelType" in model
+  );
+}
+
+function isChatModelWithBindTools(
+  llm: LanguageModelLike
+): llm is BaseChatModel & Required<Pick<BaseChatModel, "bindTools">> {
+  if (!isBaseChatModel(llm)) return false;
+  return "bindTools" in llm && typeof llm.bindTools === "function";
+}
+
+export interface ModelRequest {
+  model: LanguageModelLike;
+  systemPrompt?: string;
+  messages: BaseMessage[];
+  toolChoice: ToolChoice | undefined;
+  middleware: AgentMiddleware[] | undefined;
+  tools: Tool[];
+}
+
+export function createAgent(options: {
+  model: LanguageModelLike;
+  tools: Tool[] | ToolNode;
+  systemPrompt?: string;
+  middleware?: AgentMiddleware[];
+}) {
+  const middleware = options.middleware ?? [];
+
+  // init tool node
+  const toolNode = Array.isArray(options.tools)
+    ? new ToolNode(options.tools)
+    : options.tools;
+
+  type MiddlewareMap = Record<
+    string,
+    (
+      state: AgentState["State"]
+    ) => Promise<AgentState["Update"]> | AgentState["Update"]
+  >;
+
+  const toolNodeBefore = middleware.reduce<MiddlewareMap>(
+    (acc, { name, beforeModel }) => {
+      if (beforeModel) acc[`${name}.beforeModel`] = beforeModel;
+      return acc;
+    },
+    {}
+  );
+
+  const toolNodeAfter = middleware.reduce<MiddlewareMap>(
+    (acc, { name, afterModel }) => {
+      if (afterModel) acc[`${name}.afterModel`] = afterModel;
+      return acc;
+    },
+    {}
+  );
+
+  const builder = new StateGraph(AgentState).addNode({
+    ...toolNodeBefore,
+    ...toolNodeAfter,
+
+    modelRequest: async (state: AgentState["State"]) => {
+      let request: ModelRequest = {
+        model: options.model,
+        messages: state.messages,
+        toolChoice: unknown,
+        tools: toolNode.tools,
+        systemPrompt: options.systemPrompt,
+        middleware: options.middleware,
+      };
+
+      for (const middleware of options.middleware ?? []) {
+        if (middleware.modifyModelRequest != null) {
+          request = middleware.modifyModelRequest(request);
+        }
+      }
+
+      // prepare messages
+      const messages =
+        request.systemPrompt != null
+          ? [new SystemMessage(request.systemPrompt), ...request.messages]
+          : request.messages;
+
+      // call model
+      const model = isChatModelWithBindTools(request.model)
+        ? request.model.bindTools(request.tools, {
+            tool_choice: request.toolChoice,
+          })
+        : request.model;
+
+      // TODO: add support for responseFormat
+      const output = await model.invoke(messages);
+      return { messages: output, response: undefined };
+    },
+
+    tools: toolNode,
+  });
+
+  const firstNode = Object.keys(toolNodeBefore).at(0) ?? "modelRequest";
+  const lastNode = Object.keys(toolNodeAfter).at(0) ?? "modelRequest";
```

Comment:
Given `afterModel` hooks run in reversed order, shouldn't this be
```suggestion
  const lastNode = Object.keys(toolNodeAfter).at(-1) ?? "modelRequest";
```

Reply:
This should be correct, since toolNodeAfter callbacks are inserted in the same order as toolNodeBefore